imap_processing/swe/l1a/swe_science.py

```diff
@@ -0,0 +1,174 @@
+    # NOTE: for (step_size[index] - 1) / 2, we only keep the whole number part of
+    # the quotient
+
+    return (
+        uncompress_table[index]["base"]
+        + (multi * uncompress_table[index]["step_size"])
+        + ((uncompress_table[index]["step_size"] - 1) // 2)
+    )
+
+
+def add_metadata_to_array(data_packet, metadata_arrays):
+    """This function add metadata to metadata_arrays.
+
+    Parameters
+    ----------
+    data_packet : space_packet_parser.parser.Packet
+        SWE data packet
+    metadata_arrays : dict
+        metadata arrays
+    """
+    for key, value in data_packet.header.items():
+        metadata_arrays[key].append(value.raw_value)
+
+    for key, value in data_packet.data.items():
+        if key != "SCIENCE_DATA":
+            metadata_arrays[key].append(value.raw_value)
+
+
+def init_metadata_array(data_packet):
+    """This function read metadata from data packet.
+
+    Parameters
+    ----------
+    data_packet : space_packet_parser.parser.Packet
+        SWE data packet
+
+    Returns
+    -------
+    dict
+        metadata
+    """
+    metadata = {}
+    for key, _ in data_packet.header.items():
+        metadata[key] = []
+
+    for key, _ in data_packet.data.items():
+        if key != "SCIENCE_DATA":
+            metadata[key] = []
+
+    return metadata
+
+
+def swe_science(decom_data):
+    """SWE L1A algorithm steps:
+        - Read data from each SWE packet file
+        - Uncompress counts data
+        - Store metadata fields and data in DataArray of xarray
+        - Save data to dataset
+
+    In each packet, SWE collects data for 15 seconds. In each second,
+    it collect data for 12 energy steps and at each energy step,
+    it collects 7 data from each 7 CEMs.
+
+    Each L1A data from each packet will have this shape: 15 rows, 12 columns,
+    and each cell in 15 x 12 table contains 7 element array.
+    These dimension maps to this:
+        15 rows --> 15 seconds
+        12 column --> 12 energy steps in each second
+        7 element --> 7 CEMs counts
+
+    In L1A, we don't do anything besides read raw data, uncompress counts data and
+    store data in 15 x 12 x 7 array.
+
+    SWE want to keep all value as it is in L1A. Post L1A, we group data into full cycle
+    and convert raw data to engineering data as needed.
+
+    Parameters
+    ----------
+    packet_file : str
+        packet file path
+
+    Returns
+    -------
+    xarray.Dataset
+        xarray dataset with data.
+    """
+    science_array = []
+
+    metadata_arrays = init_metadata_array(decom_data[0])
+
+    for data_packet in decom_data:
+        # read raw data
+        binary_data = data_packet.data["SCIENCE_DATA"].raw_value
+        # read raw data as binary array using bitstring
+        # Eg. "0b0101010111"
+        bit_array = bitstring.ConstBitStream(bin=binary_data)
+        # chunk bit array into 1260 units, each with 8-bits
+        byte_data = np.frombuffer(bit_array.bytes, dtype=np.uint8)
```

Comment:
```suggestion
        # read raw data
        binary_data = data_packet.data["SCIENCE_DATA"].raw_value
        # read binary string as an int and then convert it to
        # bytes.
        # Eg. "0000000011110011" --> b'\x00\xf3'
        byte_data = int(binary_data, 2).to_bytes(1260, byteorder='big')
        # convert bytes to numpy array of uint8
        raw_counts = np.frombuffer(byte_data, dtype=np.uint8)
```
@greglucas Does this change look ok?

Reply:
Yeah. I personally think that looks better and then you can remove the explicit dependency in our requirements. But looks good either way to me!

Reply:
Perfect. I can remove import in mine but I can't remove `bitstring` from dependency because IDEX code is using it.